### PR TITLE
feat(opensphere): add support to convert implicit require()'s

### PR DESCRIPTION
**Pull request by glennkitchellcaci**
Provides configuration point for replacing modules/provides that are assumed to be available; e.g. `os.settings`.get(...). Supports non-matching requires (e.g. require('os.config.Settings'); and/or singletons using getInstance().

To get more configs into the implicit require replacements list, create a file ../my-project-directory/.jscodeshift.json with the format:
`{"globals": { "my.global.namespace": {"require": "my.alternate.namespace", "singleton": false}}}` where "require" defaults to match the key when undefined and "singleton" defaults to false when undefined.

Potential future improvements include:
- Checking the number of .getInstance() calls for one replacement in a function and creating a local const when more than 2
 - Checking the number of .getInstance() calls for one replacement in a file, and creating a file const when more than 10

### CODE REVIEW

File: src/utils/goog.js [hunk 1]
```diff
@@ -313,6 +315,9 @@ const replaceLegacyRequire = (root, toReplace) => {
     }
   }
 
+  // add .getInstance() if needed for the in-line replacements
+  if (singleton === true) varName = `${varName}.getInstance()`;
```
Comment (schmidtk) [inline src/utils/goog.js:319]: This is not the proper way to form the AST node for what you're doing here. The MemberExpression should be replaced with an ExpressionStatement that calls `getInstance` from `varName`. You can see the AST tree if you go to [ASTExplorer](https://astexplorer.net/), change the parser to `esprima`, and put this as the source:

```
alertManager.getInstance()
```

File: src/transforms/es6/providetomodule.js [hunk 2]
```diff
@@ -6,6 +6,22 @@ const {addRequire, isClosureClass, isControllerClass, isDirective, isGoogDefine,
...
+    try {
+      const config = JSON.parse(fs.readFileSync(`${workspace}${folder}/${filename}`, 'utf8'));
+      if (config) {
+        addOSGlobals(config['globals']);
+      }
+    } catch (e) {
+      // no file found or couldn't read it
+    }
+  });
+})(require('fs'), '../', '.jscodeshift.json');
```
Comment (schmidtk) [inline src/transforms/es6/providetomodule.js:24]: Typically all `require` statements should be at the top of the file. Is there a reason this has to be a param for your IIFE?

File: src/transforms/es6/providetomodule.js [hunk 3]
```diff
@@ -6,6 +6,22 @@ const {addRequire, isClosureClass, isControllerClass, isDirective, isGoogDefine,
...
+// on creation, read ".jscodeshift.json" file(s) from this and local folders
+(function(fs, workspace, filename){
+  fs.readdirSync(workspace).forEach((folder) => {
+    try {
+      const config = JSON.parse(fs.readFileSync(`${workspace}${folder}/${filename}`, 'utf8'));
+      if (config) {
+        addOSGlobals(config['globals']);
+      }
+    } catch (e) {
+      // no file found or couldn't read it
```
Comment (schmidtk) [inline src/transforms/es6/providetomodule.js:21]: `fs.existsSync` can be used to check if the file exists before trying to read it. If it exists and the read fails, an error should be thrown so the script exits and the developer knows their config file is invalid.

**Review by schmidtk — CHANGES_REQUESTED** [review]

File: src/utils/goog.js (hunk 1 above)
Comment (glennkitchellcaci) [reply src/utils/goog.js]: Booya -- now I'm using the memberexpressions so it looks right in the AST.

File: src/transforms/es6/providetomodule.js (hunk 2 above)
Comment (glennkitchellcaci) [reply src/transforms/es6/providetomodule.js]: Moved!

File: src/transforms/es6/providetomodule.js (hunk 3 above)
Comment (glennkitchellcaci) [reply src/transforms/es6/providetomodule.js]: Ah, much better. Ty ty.

**Review by schmidtk — APPROVED** [review]

**Pull request merged**